src/libraries/System.Text.Json/source_generation/src/JsonSourceGeneratorHelper.Generate.cs

```diff
@@ -0,0 +1,672 @@
+            foreach (TypeMetadata typeMetadata in rootTypeMetadata)
+            {
+                if (typeMetadata.ClassType != ClassType.TypeUnsupportedBySourceGen)
+                {
+                    sb.Append($@"
+            if (type == typeof({typeMetadata.Type.GetUniqueCompilableTypeName()}))
+            {{
+                return this.{typeMetadata.FriendlyName};
+            }}
+");
+                }
+            }
+
+            sb.Append(@"
+            return null!;
+        }
+    }
+}
+");
+
+            return sb.ToString();
+        }
+
+        private static string GetUsingStatementsString(TypeMetadata typeMetadata)
+        {
+            HashSet<string> usingStatements = GetUsingStatements(typeMetadata);
+            return GetUsingStatementsString(usingStatements);
+        }
+
+        private static string GetUsingStatementsString(HashSet<string> usingStatements)
+        {
+            string[] usingsArr = usingStatements.ToArray();
+            Array.Sort(usingsArr);
+            return string.Join("\n", usingsArr);
+        }
+
+        private static HashSet<string> GetUsingStatements(TypeMetadata typeMetadata)
+        {
+            HashSet<string> usingStatements = new();
+
+            // Add library usings.
+            usingStatements.Add(FormatAsUsingStatement("System.Runtime.CompilerServices"));
+            usingStatements.Add(FormatAsUsingStatement("System.Text.Json"));
+            usingStatements.Add(FormatAsUsingStatement("System.Text.Json.Serialization"));
+            usingStatements.Add(FormatAsUsingStatement("System.Text.Json.Serialization.Metadata"));
+
+            // Add imports to root type.
+            usingStatements.Add(FormatAsUsingStatement(typeMetadata.Type.Namespace));
+
+            switch (typeMetadata.ClassType)
+            {
+                case ClassType.Nullable:
+                    {
+                        AddUsingStatementsForType(typeMetadata.NullableUnderlyingTypeMetadata!);
+                    }
+                    break;
+                case ClassType.Enumerable:
+                    {
+                        AddUsingStatementsForType(typeMetadata.CollectionValueTypeMetadata);
+                    }
+                    break;
+                case ClassType.Dictionary:
+                    {
+                        AddUsingStatementsForType(typeMetadata.CollectionKeyTypeMetadata);
+                        AddUsingStatementsForType(typeMetadata.CollectionValueTypeMetadata);
+                    }
+                    break;
+                case ClassType.Object:
+                    {
+                        if (typeMetadata.PropertiesMetadata != null)
+                        {
+                            foreach (PropertyMetadata metadata in typeMetadata.PropertiesMetadata)
+                            {
+                                AddUsingStatementsForType(metadata.TypeMetadata);
```

Comment:
Can this result in duplicated `using` statements?

Reply:
It's adding to a hash set so there won't be duplicates.